securedrop_client/storage.py

```diff
@@ -226,37 +227,37 @@ def __update_submissions(model: Union[Type[File], Type[Message]],
     * Local submissions not returned in the remote submissions are deleted
       from the local database.
     """
-    local_uuids = {submission.uuid for submission in local_submissions}
-    for submission in remote_submissions:
-        if submission.uuid in local_uuids:
-            local_submission = [s for s in local_submissions
-                                if s.uuid == submission.uuid][0]
-
-            lazy_setattr(local_submission, "size", submission.size)
-            lazy_setattr(local_submission, "is_read", submission.is_read)
-            lazy_setattr(local_submission, "download_url", submission.download_url)
-
-            # Removing the UUID from local_uuids ensures this record won't be
-            # deleted at the end of this function.
-            local_uuids.remove(submission.uuid)
-            logger.debug('Updated submission {}'.format(submission.uuid))
-        else:
-            # A new submission to be added to the database.
-            _, source_uuid = submission.source_url.rsplit('/', 1)
-            source = session.query(Source).filter_by(uuid=source_uuid).first()
-            if source:
-                ns = model(source_id=source.id, uuid=submission.uuid, size=submission.size,
-                           filename=submission.filename, download_url=submission.download_url)
-                session.add(ns)
-                logger.debug('Added new submission {}'.format(submission.uuid))
+    local_submissions_by_uuid = {s.uuid: s for s in local_submissions}
+    remote_submissions_by_source: Dict[str, list] = collections.defaultdict(list)
+    for s in remote_submissions:
+        remote_submissions_by_source[s.source_uuid].append(s)
+
+    for source_uuid, submissions in remote_submissions_by_source.items():
+        source = session.query(Source).filter_by(uuid=source_uuid).first()
```

Comment:
before we would look up the source right before adding a new submission to the database. now looks like we look up the source for every unique source we get back from the server, right? so this piece of the change should be theoretically faster if there are many new submissions for one source, but should be slower if there are many more sources and not a lot of new submissions, right?

Reply:
Again, good call. I've added the source cache and reverted to just iterating the remote submissions/replies.